### ISSUE TITLE
ZerO for convolution layers

### ISSUE THREAD

**Issue by AlgebraLoveme** [jiaweizzhao/ZerO-initialization#1]
Dear authors,

Nice work! Congratulations.

So far, the repository only contains ZerO for linear layers. Could you please update the ZerO for convolutional layers as well (algorithm 2)? I know they have much similarity, but we just want to make sure we don't misunderstand anything about the pseudocode.

Thanks

**Comment by PiotrCzapla** [comment]
Hi, I'm trying to reproduce your paper. I've managed to get the linear example working, but I struggle with convolutions.

The algorithm2 in the paper is a bit unclear regarding the use of n (why is it k//2), and what is the value of `m` and `c`, (the text looks like it is copied from algorithm 1). 

After second read I think I've got it, and I managed to get the network to train, although the training. Although the results are so so on a simple conv network, probably  I need to experiment the LR warmup.

My assumption now is that we initialise the kernel to zeros and set the middle point to Hadamard matrix.

The current implementation looks as follows:

```py
@torch.no_grad()
def init_identity_(w):
    if w.dim() == 2:
        ## nn.init.eye_ doesn't work on mps for tensor views 
        w.copy_(torch.eye(*w.shape, dtype=w.dtype))
    elif w.dim() == 4:
        n = int(w.shape[-1]/2)
        nn.init.zeros_(w)
        init_identity_(w[...,n,n])
        #w.copy_(torch.eye(*w.shape[2:]))
    else:
        raise NotImplemented('not impl yet')
    return w

def hadamard(rank, dtype=None):
    H = torch.tensor([[1.]], dtype=dtype)
    for i in range(0, rank): H =  torch.cat([torch.cat([H, H]), torch.cat([H, -H])], dim=1)
    return H

@torch.no_grad()
def init_hadamard_(w):
    if w.dim() == 2: 
        m,n = w.shape
        rank = m.bit_length()
        H = hadamard(rank)
        w.copy_(torch.eye(m, H.shape[0]) @ (H/(2**((rank-1)/2))) @ torch.eye(H.shape[0], n)) 
    elif w.dim() == 4:
        init.zeros_(w)
        n = int(w.shape[-1]/2)
        init_hadamard_(w[..., n, n])
        #w.copy_(init_hadamard_(torch.zeros(*w.shape[2:])).expand_as(w)) # old incorrect read of the algo2 pseudo code
    return w

def init_zero_(w):
    if w.dim() in [2,4]: # lin: [out, in] & conv [out, in, k, k] 
        cout,cin, *_ = w.shape
        if cin >= cout: init_identity_(w)
        else: init_hadamard_(w)
    else:
        raise ValueError(f'Initialization of weights of shape {w.shape} is not supported dim() has to be 2 or 4')
    return w

**Comment by snoop2head** [comment]
Below is my personal implementation in pytorch! 
Any feedbacks will be helpful for me, if possible 🤗 @jiaweizzhao 

```python
def init_ZerO_linear(matrix_tensor):
    with torch.no_grad(): 
        m = matrix_tensor.size(0)
        n = matrix_tensor.size(1)

        if m <= n:
            init_matrix = torch.nn.init.eye_(torch.empty(m, n))
        elif m > n:
            clog_m = math.ceil(math.log2(m))
            p = 2**(clog_m)
            init_matrix = torch.nn.init.eye_(torch.empty(m, p)) @ (torch.tensor(hadamard(p)).float()/(2**(clog_m/2))) @ torch.nn.init.eye_(torch.empty(p, n))

        return nn.Parameter(init_matrix)

def init_ZerO_convolution(convolution_tensor):
    with torch.no_grad():
        m = convolution_tensor.size(0)
        n = convolution_tensor.size(1)
        k = convolution_tensor.size(2)
        l = convolution_tensor.size(3)

        index = int(math.floor(k/2))
        if m <= n:
            convolution_tensor[:, :, index, index] = torch.nn.init.eye_(torch.empty(m, n))

        elif m > n:
            clog_m = math.ceil(math.log2(m))
            p = 2**(clog_m)
            convolution_tensor[:, :, index, index] = torch.nn.init.eye_(torch.empty(m, p)) @ (torch.tensor(hadamard(p)).float()/(2**(clog_m/2))) @ torch.nn.init.eye_(torch.empty(p, n))

        return nn.Parameter(convolution_tensor)

def init_ZerO(module):
    # linear and conv1d weight initialization
    if isinstance(module, nn.Linear) or isinstance(module, nn.Conv1d):
        module.weight = init_ZerO_linear(module.weight)
        if module.bias is not None:
            nn.init.constant_(module.bias, 0)

    # conv initialization
    elif isinstance(module, nn.Conv2d):
        module.weight = init_ZerO_convolution(module.weight)
        if module.bias is not None:
            nn.init.constant_(module.bias, 0)

    # batchnorm initialization: "For ResNet with batch normalization, we follow the standard practice to initialize the scale and bias in batch normalization as one and zero"
    elif isinstance(module, nn.BatchNorm1d) or isinstance(module, nn.BatchNorm2d) or isinstance(module, nn.BatchNorm3d):
        nn.init.constant_(module.weight, 1)
        nn.init.constant_(module.bias, 0)
```
